# Problem Set 3

Author: Andre Oviedo Mendoza

# Part 1) {-}

## Problem 1 {-}


First, we will need to reorder the observations so that our first $n_0$ rows are $s_o$ and the last $n_1$ rows are $s_1$.

$\boldsymbol{S}^{\prime} \boldsymbol{S}$ will be a diagonal matrix where each element of the diagonal entry are the number of observations with $S_i = s_i$.

Therefore, 

$$\operatorname{plim}(\frac{1}{n}\boldsymbol{S}^{\prime} \boldsymbol{S}) = \text{diag}(\mathbb{P}(S=s_0), \mathbb{P}(S=s_1), ..., \mathbb{P}(S=s_L))$$

For $\boldsymbol{S}^{\prime}T$, each entry is the sum of T values for observations with $S_i = s_l$, so:

$$\operatorname{plim}(\frac{1}{n}\boldsymbol{S}^{\prime}T) = (\mathbb{P}(T=1|S=s_0)\mathbb{P}(S=s_0), ..., \mathbb{P}(T=1|S=s_L)\mathbb{P}(S=s_L))'$$

Our first-stage coefficients are

$$\operatorname{plim}(\frac{1}{n}\boldsymbol{S}^{\prime} \boldsymbol{S})^{-1}\operatorname{plim}(\frac{1}{n}\boldsymbol{S}^{\prime}T) = (g(s_0), g(s_1), ..., g(s_L))'$$

We know that

$$\operatorname{plim}(\frac{1}{n}\boldsymbol{S}^{\prime} \boldsymbol{S})$$ 

is a diagonal matrix with $\mathbb{P}(S=s_l)$ on the diagonal, and

$$\operatorname{plim}(\frac{1}{n}\boldsymbol{S}^{\prime}T)$$ 

is a vector with entries $\mathbb{P}(T=1|S=s_l)\mathbb{P}(S=s_l)$

When we multiply the inverse of the first matrix by the second vector. The $\mathbb{P}(S=s_l)$ terms cancel out and we're left with just $\mathbb{P}(T=1|S=s_l)$ for each level $l$.

From the threshold crossing model, we know $\mathbb{P}(T=1|S=s_l) = g(s_l)$.

Therefore, the first-stage coefficient on each indicator $$\mathbb{1}(S_i = s_l)$$ equals $$g(s_l)$$, which is what we wanted to prove.

## Problem 2 {-}

First, write out the covariance:
$$ \operatorname{Cov}(Y,g(S)) = \mathbb{E}[Y g(S)] - \mathbb{E}[Y]\mathbb{E}[g(S)] $$

by using the law of total expectation, we know that $ \mathbb{E}[Y g(S)] = \sum_{l=0}^{L} \mathbb{E}[Y|S=s_l]g(s_l)\mathbb{P}(S=s_l) $


For each conditional expectation, using the fact that $g(s)$ is increasing in $s$ we can use the telescoping sum to rewrite $ \mathbb{E}[Y|S=s_l] $ as

$$ \mathbb{E}[Y|S=s_l] = \mathbb{E}[Y|S=s_0] + \sum_{m=1}^{l}(\mathbb{E}[Y|S=s_m] - \mathbb{E}[Y|S=s_{m-1}]) $$

Substituting this into our covariance formula,

$$ \operatorname{Cov}(Y,g(S)) = \sum_{l=0}^{L} \left(\mathbb{E}[Y|S=s_0] + \sum_{m=1}^{l}(\mathbb{E}[Y|S=s_m] - \mathbb{E}[Y|S=s_{m-1}])\right)g(s_l)\mathbb{P}(S=s_l) - \mathbb{E}[Y]\mathbb{E}[g(S)] $$

and after distributing terms and rearranging we arrive to our desired result:

$$ \operatorname{Cov}(Y,g(S)) = \sum_{l=0}^{L}\left(\sum_{m=1}^{l}\left[\mathbb{E}\left[Y \mid S=s_{m}\right]-\mathbb{E}\left[Y \mid S=s_{m-1}\right]\right]\right)\left(g\left(s_{l}\right)-\mathbb{E}[g(S)]\right) \mathbb{P}\left(S=s_{l}\right) $$


## Problem 3 {-}

We can start fro the result from Problem 2

$$ \operatorname{Cov}(Y,g(S)) = \sum_{l=0}^{L}\left(\sum_{m=1}^{l}\left[\mathbb{E}\left[Y \mid S=s_{m}\right]-\mathbb{E}\left[Y \mid S=s_{m-1}\right]\right]\right)\left(g\left(s_{l}\right)-\mathbb{E}[g(S)]\right) \mathbb{P}\left(S=s_{l}\right) $$

Following Chamberlain's notes, we know that for any $s$ 

$$ \mathbb{E}[Y|S=s] = \mathbb{E}[Y(0)] + g(s)\theta + \int_0^{g(s)} \mathbb{E}[U_1-U_0|V=v]dv $$

Therefore

$$ \mathbb{E}[Y|S=s_m] - \mathbb{E}[Y|S=s_{m-1}] = [g(s_m) - g(s_{m-1})]\mathbb{E}[Y(1)-Y(0)|g(s_{m-1}) \leq V \leq g(s_m)] $$

From the defintion of the $\operatorname{LATE}_{m-1}^m$ we know that

$$ \operatorname{LATE}_{m-1}^m = \mathbb{E}[Y(1)-Y(0)|g(s_{m-1}) \leq V \leq g(s_m)] $$

So, rearranging the terms in our covariance formula and replacing the $\operatorname{LATE}_{m-1}^m$ we get

$$ \operatorname{Cov}(Y,g(S)) = \sum_{l=0}^{L}\left(\sum_{m=1}^{l}[g(s_m) - g(s_{m-1})]\operatorname{LATE}_{m-1}^m\right)\left(g\left(s_{l}\right)-\mathbb{E}[g(S)]\right) \mathbb{P}\left(S=s_{l}\right) $$




## Problem 4 {-}

To show that both expressions are equivalent, we have to understand the LHS

In this expression

- We first sum over $l$ from 0 to $L$.
- For each $l$, we sum over $m$ from 1 to $l$.
- The term inside the summation over $m$ represents the difference in values of $g(s_m)$ and $g(s_{m-1})$, multiplied by the local average treatment effect (LATE) between these two points.
- Finally, we multiply by the difference $g(s_l) - \mathbb{E}[g(S)]$, weighted by the probability $\mathbb{P}(S = s_l)$.

The key step in this problem is to **switch the order of summation**. Instead of summing over $l$ first and then $m$, we will sum over $m$ first and then $l$. This is valid because we are just reordering how we sum the terms, which does not change their values.

This works because:
- In the original expression, for each fixed $l$, we sum over all $m$ such that $m \leq l$.
- After switching the order, for each fixed $m$, we sum over all $l$ such that $l \geq m$.

This reordering works because every term in the original double summation appears exactly once, and changing the order simply reorganizes how we group those terms.

$$
\sum_{m=1}^{L} \mathrm{LATE}_{m-1}^{m} \cdot \left( g(s_m) - g(s_{m-1}) \right) \cdot \sum_{l=m}^{L} \left( g(s_l) - \mathbb{E}[g(S)] \right) P(S = s_l)
$$


## Problem 5 {-}


From Problem 2, we can use the same approach to write Cov(T,g(S)) in terms of differences in conditional expectations:

$$ \operatorname{Cov}(T, g(S))=\sum_{l=0}^{L}\left(\sum_{m=1}^{l}\left[\mathbb{E}\left[T \mid S=s_{m}\right]-\mathbb{E}\left[T \mid S=s_{m-1}\right]\right]\right)\left(g\left(s_{l}\right)-\mathbb{E}[g(S)]\right) \mathbb{P}\left(S=s_{l}\right) $$

From Problem 1, we know that

$$ \mathbb{E}[T|S=s_l] = g(s_l) $$

Therefore

$$ \mathbb{E}[T|S=s_m] - \mathbb{E}[T|S=s_{m-1}] = g(s_m) - g(s_{m-1}) $$

Using Problem 4's approach to change the order of summation, we get

$$ \operatorname{Cov}(T, g(S))=\sum_{m=1}^{L}\left(g\left(s_{m}\right)-g\left(s_{m-1}\right)\right) \sum_{l=m}^{L}\left(g\left(s_{l}\right)-\mathbb{E}[g(S)]\right) \mathbb{P}\left(S=s_{l}\right) $$

This proves the desired result. The covariance between $T$ and $g(S)$ can be written as a weighted sum of the differences in $g(S)$ at adjacent levels, where the weights involve the probabilities and deviations from the mean at higher levels of the instrument.

## Problem 6 {-}

We have already calculated the components of $\beta_\text{2SLS}$ in Problem 5 and 4.

$$ \beta_\text{2SLS} = \frac{\operatorname{Cov}(Y,g(S))}{\operatorname{Cov}(T,g(S))} $$

Substituting the components we found in Problems 5 and 4, we get

$$ \beta_\text{2SLS} = \frac{\sum_{m=1}^{L} \mathrm{LATE}_{m-1}^{m} \cdot \left( g(s_m) - g(s_{m-1}) \right) \cdot \sum_{l=m}^{L} \left( g(s_l) - \mathbb{E}[g(S)] \right) P(S = s_l)}{\sum_{m=1}^{L}\left(g\left(s_{m}\right)-g\left(s_{m-1}\right)\right) \sum_{l=m}^{L}\left(g\left(s_{l}\right)-\mathbb{E}[g(S)]\right) \mathbb{P}\left(S=s_{l}\right)} $$

From this equation we can extract a set of weights $w_m$ that are the coefficients on $\mathrm{LATE}_{m-1}^m$ in the 2SLS regression:

$$ w_m = \frac{(g(s_m) - g(s_{m-1})) \sum_{l=m}^{L}(g(s_l) - \mathbb{E}[g(S)])\mathbb{P}(S=s_l)}{\sum_{m=1}^{L}(g(s_m) - g(s_{m-1})) \sum_{l=m}^{L}(g(s_l) - \mathbb{E}[g(S)])\mathbb{P}(S=s_l)} $$

More importantly, we can see that the weights sum to one by construction: the numerator is an element of the (summed) denominator. Will the weights always be positve? We can extract the expected sign of each of the components of $w_m$ from the expression:

- $g(s_m) - g(s_{m-1}) > 0$ by assumption   
- The sum $\sum_{l=m}^{L}(g(s_l) - \mathbb{E}[g(S)])P(S=s_l)$ is positive for $l \geq m$ since $g(s)$ is increasing
- The denominator is positive as it's a sum of positive terms

This weighted average interpretation is desirable because:
- It shows $\beta_\text{2SLS}$ is a proper convex combination of local average treatment effects
- The positive weights ensure the estimate lies between the minimum and maximum LATEs
- Each LATE is weighted by both the size of the complier group and the instrument's strength at that level


# Part 2) {-}


## Problem 7 {-}

The instrumental variables (IV) estimator for $\beta$ solves the following moment condition:

$$
\mathbb{E}[Z(Y - D\beta)] = 0
$$

This implies that the IV estimator $\hat{\beta}$ is obtained by solving:

$$
\hat{\beta} = \left( Z'D \right)^{-1} Z'Y
$$

where $Z$ is the matrix of instruments, $D$ is the endogenous regressor, and $Y$ is the outcome variable.

To construct confidence intervals for $\hat{\beta}$, we can use **inverted hypothesis tests** based on the moment condition. The idea is to test whether a particular value of $\beta_0$ satisfies the moment condition:

$$
\mathbb{E}[Z(Y - D\beta_0)] = 0
$$

For a given value of $\beta_0$, we can compute the sample analog of this moment condition:

$$
m(\beta_0) = \frac{1}{n} Z'(Y - D\beta_0)
$$

If $m(\beta_0)$  is close to zero, then $\beta_0$ is a plausible value for $\beta$. If not, we reject $\beta_0$.

We can construct a **test statistic** based on $m(\beta_0)$. A natural choice is a quadratic form that accounts for the covariance structure of the moment conditions. Specifically, we can use a Wald-type test statistic

$$
T(\beta_0) = m(\beta_0)' W^{-1} m(\beta_0)
$$

where $W^{-1}$ is an estimate of the covariance matrix of the moment conditions. Under the null hypothesis that $\mathbb{E}[Z(Y - D\beta_0)] = 0$, this test statistic follows a chi-squared distribution with degrees of freedom equal to the number of instruments.

To construct confidence intervals for $\hat{\beta}$, we can invert this hypothesis test. Specifically, we find all values of $\beta_0$ such that the test statistic $T(\beta_0)$ does not exceed a critical value from the chi-squared distribution.

Formally, let $c_{\alpha}$ be the critical value from the chi-squared distribution at significance level $\alpha$. The **confidence interval** for $\hat{\beta}$ is given by:

$$
CI_{1-\alpha} = \left\{ \beta_0 : T(\beta_0) \leq c_{\alpha} \right\}
$$

This approach ensures that we include all values of $\beta_0$ that are consistent with the moment condition within a certain confidence level.

In practice, you would:
1. Estimate $m(\beta_0)$ for different candidate values of $\beta_0$.
2. Compute the test statistic $T(\beta_0)$.
3. Compare $T(\beta_0)$ to the critical value from a chi-squared distribution.
4. Construct confidence intervals by finding all values of $\beta_0$ where $T(\beta_0)$ does not exceed this critical value.

# Part 3) {-}

## Problem 8 {-}

In [68]:
# Source the Python script
exec(open('dml_example.py').read())

## Problem 8 {-}

In [70]:
# Using the variables from the data:

# y = "logghomr" (log gun homicide rate)
# d = "logfssl" (log gun ownership proxy)

import statsmodels.api as sm

simple_model = sm.OLS(usedata['logghomr'], sm.add_constant(usedata['logfssl'])).fit()
print(simple_model.summary())

                            OLS Regression Results                            
Dep. Variable:               logghomr   R-squared:                       0.111
Model:                            OLS   Adj. R-squared:                  0.111
Method:                 Least Squares   F-statistic:                     487.7
Date:                Tue, 19 Nov 2024   Prob (F-statistic):          6.26e-102
Time:                        00:16:25   Log-Likelihood:                -5661.6
No. Observations:                3900   AIC:                         1.133e+04
Df Residuals:                    3898   BIC:                         1.134e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.3288      0.044      7.480      0.0

In [71]:
# This gives similar coef estimates as the code from the book (which fixes the std err)

# Baseline OLS
X = sm.add_constant(usedata['logfssl'])
y = usedata['logghomr']
lm0 = sm.OLS(y, X).fit(cov_type='HC3')
vc0 = lm0.cov_params()
coef = lm0.params['logfssl']
std_err = np.sqrt(vc0.loc['logfssl', 'logfssl'])
print("Baseline OLS:", coef, "(", std_err, ")")

Baseline OLS: 0.30202337690461406 ( 0.012565715795207547 )


For this to be interpreted as a causal effect, we would need several key identification assumptions:


1. **Exogeneity/Unconfoundedness**: Gun ownership levels must be *as good as* randomly assigned, meaning there are no unmeasured factors that affect both gun ownership and homicide rates. This is a strong assumption because:
   - Areas with higher crime might have more people buying guns for protection
   - Cultural and socioeconomic factors likely influence both gun ownership and violence
   - Local policies could affect both variables  
2. **Stable Unit Treatment Value Assumption (SUTVA)**:
   - One area's gun ownership levels shouldn't affect homicide rates in other areas
   - This might be violated due to cross-border gun trafficking or spillover effects
3. **No Reverse Causality**:
   - Changes in homicide rates shouldn't cause changes in gun ownership
   - This is likely violated as people may buy guns in response to crime

## Problem 9 {-}

In [72]:
# Create formula with all control variables
controls = [d] + X1 + X2 + census  # Combining all control variables
X = usedata[controls]
X = sm.add_constant(X)  # Add constant term
X['logfssl'] = usedata['logfssl']  # Add treatment variable

controlled_model = sm.OLS(usedata['logghomr'], X).fit()
#print(controlled_model.summary()) Not printing cause of length

In [73]:
# Regression on baseline controls from the book, again, similar answer
varlist = [d] + X1 + X2 + census
X = sm.add_constant(usedata[varlist])
y = usedata['logghomr']
lmC = sm.OLS(y, X).fit(cov_type='HC3')
vcC = lmC.cov_params()
coef = lmC.params['logfssl']
std_err = np.sqrt(vcC.loc['logfssl', 'logfssl'])
print("OLS with Controls:", coef, "(", std_err, ")")

OLS with Controls: 0.3102764669361729 ( 0.0729464029073983 )


In [74]:
# We can also follow the book and add all controls
X = sm.add_constant(usedata.drop(columns=['logghomr'])) 
controlled_model = sm.OLS(usedata['logghomr'], X).fit()
#print(controlled_model.summary()) Not printing cause of length

In [75]:
# From the book

X = sm.add_constant(usedata.drop(columns=['logghomr']))
y = usedata['logghomr']
lmA = sm.OLS(y, X).fit(cov_type='HC3')
vcA = lmA.cov_params()
coef = lmA.params['logfssl']
std_err = np.sqrt(vcA.loc['logfssl', 'logfssl'])
print("OLS All:", coef, "(", std_err, ")")

OLS All: 0.17931590182937693 ( 0.07981180505737272 )


Again, for this controlled effect to be interpreted as causal, we still need a lot of assumptions for this to be considered a causal effect. 

The main difficulty that this could have is the exogeneity condition. Gun ownership levels must still be as a good as randomly assigned, even after controlling for all other variables. This could be a really hard condition to fulfill:

- Areas with higher crime might have more people buying guns for protection
- Cultural and socioeconomic factors likely influence both gun ownership and violence
- Local policies could affect both variables

## Problem 10 {-}

In [37]:
def dml(X, D, y, modely, modeld, *, nfolds, classifier=False):
    '''
    DML for the Partially Linear Model setting with cross-fitting

    Input
    -----
    X: the controls
    D: the treatment
    y: the outcome
    modely: the ML model for predicting the outcome y
    modeld: the ML model for predicting the treatment D
    nfolds: the number of folds in cross-fitting
    classifier: bool, whether the modeld is a classifier or a regressor

    time: array of time indices, eg [0,1,...,T-1,0,1,...,T-1,...,0,1,...,T-1]
    clu: array of cluster indices, eg [1073, 1073, 1073, ..., 5055, 5055, 5055, 5055]
    cluster: bool, whether to use clustered standard errors

    Output
    ------
    point: the point estimate of the treatment effect of D on y
    stderr: the standard error of the treatment effect
    yhat: the cross-fitted predictions for the outcome y
    Dhat: the cross-fitted predictions for the treatment D
    resy: the outcome residuals
    resD: the treatment residuals
    epsilon: the final residual-on-residual OLS regression residual
    '''
    cv = KFold(n_splits=nfolds, shuffle=True, random_state=123)  # shuffled k-folds
    yhat = cross_val_predict(modely, X, y, cv=cv, n_jobs=-1)  # out-of-fold predictions for y
    # out-of-fold predictions for D
    # use predict or predict_proba dependent on classifier or regressor for D
    if classifier:
        Dhat = cross_val_predict(modeld, X, D, cv=cv, method='predict_proba', n_jobs=-1)[:, 1]
    else:
        Dhat = cross_val_predict(modeld, X, D, cv=cv, n_jobs=-1)
    # calculate outcome and treatment residuals
    resy = y - yhat
    resD = D - Dhat

    dml_data = pd.concat([pd.Series(resy, name='resy'), pd.Series(resD, name='resD')], axis=1)
    ols_mod = smf.ols(formula='resy ~ 1 + resD', data=dml_data).fit()

    point = ols_mod.params[1]
    stderr = ols_mod.bse[1]
    epsilon = ols_mod.resid

    return point, stderr, yhat, Dhat, resy, resD, epsilon

def summary(point, stderr, yhat, Dhat, resy, resD, epsilon, X, D, y, *, name):
    '''
    Convenience summary function that takes the results of the DML function
    and summarizes several estimation quantities and performance metrics.
    '''
    return pd.DataFrame({'estimate': point,  # point estimate
                         'stderr': stderr,  # standard error
                         'lower': point - 1.96 * stderr,  # lower end of 95% confidence interval
                         'upper': point + 1.96 * stderr,  # upper end of 95% confidence interval
                         'rmse y': np.sqrt(np.mean(resy**2)),  # RMSE of model that predicts outcome y
                         'rmse D': np.sqrt(np.mean(resD**2))  # RMSE of model that predicts treatment D
                         }, index=[name])

In [51]:
table = pd.DataFrame()

In [52]:
from sklearn.neural_network import MLPRegressor

# Try different NN architectures
architectures = [
    (1, 32),  # 1 layer with 32 units
    #(1, 64),  # 1 layer with 64 units 
    (2, 32),  # 2 layers with 32 units each
    (2, 64),  # 2 layers with 64 units each
    (3, 32),  # 3 layers with 32 units each
    #(3, 64)   # 3 layers with 64 units each
]

results = []
for layers, units in architectures:
    # Create hidden layer sizes tuple based on number of layers and units
    hidden_layer_sizes = tuple([units] * layers)
    
    # Create models
    modely = make_pipeline(
        StandardScaler(),
        MLPRegressor(
            hidden_layer_sizes=hidden_layer_sizes,
            max_iter=50,
            batch_size=32,
            verbose=0
        )
    )
    modeld = make_pipeline(
        StandardScaler(),
        MLPRegressor(
            hidden_layer_sizes=hidden_layer_sizes, 
            max_iter=50,
            batch_size=32,
            verbose=0
        )
    )
    
    # Run DML model
    result = dml(Z, D, Y, modely, modeld, nfolds=5, classifier=False)
    
    # Add to summary table
    name = f'NN {layers} layers {units} units'
    table = pd.concat([table, summary(*result, Z, D, y, name=name)])

table


estimate    stderr     lower     upper    rmse y  \
NN 1 layers 32 units  0.189524  0.046316  0.098744  0.280304  0.505605   
NN 2 layers 32 units  0.092973  0.042845  0.008997  0.176950  0.500504   
NN 2 layers 64 units  0.229418  0.039138  0.152707  0.306128  0.504359   
NN 3 layers 32 units -0.079973  0.032738 -0.144140 -0.015806  0.480433   

                        rmse D  
NN 1 layers 32 units  0.174429  
NN 2 layers 32 units  0.186808  
NN 2 layers 64 units  0.208103  
NN 3 layers 32 units  0.247671

In [62]:
rmses = table.iloc[:, -2:]
print("Lowest RMSE y: ", rmses.iloc[:, 0].idxmin())
print("Lowest RMSE D: ", rmses.iloc[:, 1].idxmin())

Lowest RMSE y:  NN 3 layers 32 units
Lowest RMSE D:  NN 1 layers 32 units


From the results, it seems like the NN with one layer and 32 units (number of neurons in each layer)

## Problem 11 {-}

In [63]:
# Using the NN with 1 layer and 32 units
modely = make_pipeline(
    StandardScaler(),
    MLPRegressor(
        hidden_layer_sizes=(32,),
        max_iter=50,
        batch_size=32,
        verbose=0
    )
)
modeld = make_pipeline(
    StandardScaler(), 
    MLPRegressor(
        hidden_layer_sizes=(32,),
        max_iter=50,
        batch_size=32,
        verbose=0
    )
)

# Run DML model
result = dml(Z, D, Y, modely, modeld, nfolds=5, classifier=False)

# Add to summary table
name = 'NN 1 layer 32 units'
table = pd.concat([table, summary(*result, Z, D, Y, name=name)])

print("\nDML Estimate:", result[0])



DML Estimate: 0.16428903199745787


## Problem 13 {-}

In [55]:
# Importing from the ps2

data_qr = pd.read_csv("./data_qr.csv")
data_qr.head()

birthweight  boy  married  black  age  highschool  somecollege  college  \
0         2926    0        1      0   25           1            0        0   
1         3595    0        0      1   17           0            0        0   
2         4338    1        1      0   21           0            0        0   
3         2807    1        1      0   32           0            1        0   
4         2840    1        1      0   29           1            0        0   

   prenone  presecond  prethird  smoker  cigsdaily  weightgain  
0        0          0         0       0          3          22  
1        0          0         0       1          0          44  
2        0          0         0       1          0          61  
3        0          0         0       1          0          40  
4        0          0         0       0         20          24

In [59]:
# 1. First, let's prepare the data by adding squared terms
data_qr['age2'] = data_qr['age'] ** 2
data_qr['weightgain2'] = data_qr['weightgain'] ** 2

# 2. Simple uncontrolled regression of birthweight on smoking
simple_model = sm.OLS(data_qr['birthweight'], sm.add_constant(data_qr['smoker'])).fit()
print("Uncontrolled Effect of Smoking on Birthweight:")
print(simple_model.summary().tables[1])

# 3. Prepare features for neural networks
control_vars = data_qr.drop(columns=['birthweight']).columns

control_vars

Uncontrolled Effect of Smoking on Birthweight:
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3070.7236     17.864    171.894      0.000    3035.706    3105.741
smoker       224.0707     19.128     11.715      0.000     186.577     261.565


Index(['boy', 'married', 'black', 'age', 'highschool', 'somecollege',
       'college', 'prenone', 'presecond', 'prethird', 'smoker', 'cigsdaily',
       'weightgain', 'age2', 'weightgain2'],
      dtype='object')

In [64]:
X = data_qr[control_vars].values
y = data_qr['birthweight'].values
d = data_qr['smoker'].values

# Create models using scikit-learn's MLPRegressor
modely = make_pipeline(
    StandardScaler(),
    MLPRegressor(
        hidden_layer_sizes=(64, 32),  # Two layers: 64 units, then 32 units
        max_iter=50,
        batch_size=32,
        activation='relu',
        verbose=0
    )
)

modeld = make_pipeline(
    StandardScaler(),
    MLPRegressor(
        hidden_layer_sizes=(64, 32),
        max_iter=50,
        batch_size=32,
        activation='relu',
        verbose=0
    )
)

# Run DML model
result = dml(X, d, y, modely, modeld, nfolds=5, classifier=False)

# Add to summary table
name = 'NN 2 layers 64,32 units'
table = pd.concat([table, summary(*result, X, d, y, name=name)])

print("\nDML Estimate:", result[0])


DML Estimate: -503.3289201204332


Our Double Machine Learning (DML) estimate suggests that maternal smoking during pregnancy has a negative causal effect on infant birthweight. Specifically, we find that smoking is associated with approximately 503 grams lower birthweight, which aligns with existing medical literature. This negative relationship makes intuitive sense from a biological perspective: smoking restricts blood flow and oxygen to the developing fetus, which can impair growth and development. 

This is in stark contrast to our initial naive estimate which showed a positive relationship between smoking and birthweight. The difference between these estimates highlights the importance of controlling for confounding variables through methods like DML. The naive estimate was likely biased by confounders - factors that affect both smoking behavior and birthweight - whereas DML helps isolate the true causal effect by accounting for these confounding relationships. 
